urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/MuxServer.java

```diff
@@ -0,0 +1,566 @@
+		this.right = (right == null ? new HashMap<Integer, InboundAddress>() : //
+				Collections.singletonMap(Integer.valueOf(0), right));
+	}
+
+	void addRight(final InboundAddress right) {
+		this.right.put(right.getOpts().getTunID(), right);
+	}
+
+	void listenLocal() throws IOException { // Entry Point
+		localListen = new MuxServerListenLocal(left); // Local is MUX
+		context.addReloadableAwaiter(localListen);
+		context.submitTask(localListen, "MuxInListenLeft[" + left + "]", ClientId.newId());
+	}
+
+	void listenRemote() throws IOException {
+		for (InboundAddress right : this.right.values()) {
+			try {
+				final MuxServerListenRemote remoteListen = new MuxServerListenRemote(right); // Remote is RAW
+				this.remoteListen.add(remoteListen);
+				context.addReloadableAwaiter(remoteListen);
+				context.submitTask(remoteListen, "MuxInListenRight[" + left + "|" + right + "]",
+						ClientId.newId());
+			} catch (IOException e) {
+				Log.error(this.getClass().getSimpleName() + "::listenRemote " + e.toString(), e);
+			}
+		}
+	}
+
+	void closeRemote() {
+		for (final MuxServerListenRemote remoteListen : this.remoteListen) {
+			remoteListen.setShutdown();
+		}
+	}
+
+	void closeRemote(final int id) {
+		// Send FIN
+		try {
+			final MuxPacket mux = context.allocateMuxPacket();
+			mux.fin(id);
+			local.sendLocal(mux);
+			context.releaseMuxPacket(mux);
+		} catch (Exception ign) {
+		}
+		//
+		synchronized (mapRemotes) {
+			final MuxServerRemote remote = mapRemotes.remove(id);
+			if (remote != null) {
+				remote.setShutdown();
+			}
+		}
+	}
+
+	void sendACK(final RawPacket msg) {
+		// Send ACK
+		try {
+			final MuxPacket mux = context.allocateMuxPacket();
+			mux.ack(msg.getIdChannel(), msg.getBufferLen());
+			local.sendLocal(mux);
+			context.releaseMuxPacket(mux);
+		} catch (Exception ign) {
+		}
+	}
+
+	void sendNOP() {
+		// Send NOP
+		try {
+			final MuxPacket mux = context.allocateMuxPacket();
+			mux.nop(0);
+			local.sendLocal(mux);
+			context.releaseMuxPacket(mux);
+		} catch (Exception ign) {
+		}
+	}
+
+	MuxServerRemote getRemote(final int id) {
+		synchronized (mapRemotes) {
+			return mapRemotes.get(id);
+		}
+	}
+
+	static void doSleep(final long time) {
+		try {
+			Thread.sleep(time);
+		} catch (InterruptedException ie) {
+			Thread.currentThread().interrupt();
+		}
+	}
+
+	// ============================================
+
+	class MuxServerMessageRouter {
+		void onReceiveFromLocal(final MuxServerLocal local, final MuxPacket msg) { // Local is MUX
+			// Log.debug(this.getClass().getSimpleName() + "::onReceiveFromLocal " + msg);
+			if (msg.syn()) { // This is SYN/ACK
+				final MuxServerRemote remote = getRemote(msg.getIdChannel());
+				if (remote != null)
+					remote.unlock(Constants.BUFFER_LEN * Constants.IO_BUFFERS);
+			} else if (msg.fin()) { // End SubChannel
+				Log.info(this.getClass().getSimpleName() + "::onReceiveFromLocal " + msg);
+				final MuxServerRemote remote = getRemote(msg.getIdChannel());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *remote*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).